File: solutions/scala/066-construct-binary-tree-from-preorder-and-inorder-traversal.scala
/**
 * Problem 66: Construct Binary Tree from Preorder and Inorder Traversal (LeetCode 105)
 * Difficulty: Med
 * Language: Scala
 */
def buildTree[A](preorder: List[A], inorder: List[A]): Tree[A] =
  preorder match {
    case Nil => Nil
    case r :: pre =>
      val (leftIn, _ :: rightIn) = inorder.span(_ != r): @unchecked
      val n = leftIn.length
      Node(r, buildTree(pre.take(n), leftIn),
              buildTree(pre.drop(n), rightIn))
  }

File: solutions/haskell/066-construct-binary-tree-from-preorder-and-inorder-traversal.hs
{-
  Problem 66: Construct Binary Tree from Preorder and Inorder Traversal (LeetCode 105)
  Difficulty: Med
  Language: Haskell
-}
buildTree :: Eq a => [a] -> [a] -> Tree a
buildTree [] _ = Nil
buildTree (r:pre) ino =
  let (left, _:right) = break (== r) ino
      n = length left
  in Node r (buildTree (take n pre) left)
            (buildTree (drop n pre) right)
